src/domain/lattice.ml

```diff
@@ -27,16 +27,28 @@ sig
   val pretty_diff: unit -> (t * t) -> Pretty.doc
 end
 
-(* complete lattice *)
-module type S =
+module type Bot =
 sig
-  include PO
+  type t
   val bot: unit -> t
   val is_bot: t -> bool
+end
+
+module type Top =
+sig
+  type t
   val top: unit -> t
   val is_top: t -> bool
 end
```

Comment:
Could we go for `BoundedMeetSemiLattice` and ` BoundedJoinSemiLattice` here as meaningful mathematical objects? And use this opportunity to actually make our PO a PartialOrder and not some sort of lattice in disguise, i.e., remove `join`, `meet`, `narrow` and `widen` from it?

Reply:
I could try that in #1728 to also weaken arguments of `LiftBot` and `LiftTop`.

Splitting the signatures into precise algebraic structures is nice, but it comes with a cost that we already have with `Printable` and `Lattice`: to make meaningful use of them, one needs functors for each one, e.g. `Printable.Prod`, `Lattice.POProd`, `Lattice.BoundedMeetSemiLatticeProd`, `Lattice.BoundedJoinSemiLatticeProd`, `Lattice.Prod`, etc.
Of course all code duplication can be avoided by `include`, but they still need to be separate functors with separate names.

I've thought about splitting `Printable.S` in the past, because it is even more bloated, but it just leads to a bigger mess.

Reply:
The question is whether all these lifters are used then. If we don't have a `BoundedMeetSemiLatticeProd` anywhere, no need to provide the functor for it.

Reply:
It's not completely orthogonal: I think introducing ill-defined module signatures such as `Top` (there is a special element intended to be greater than all others and a way to check whether some element is identical to it, but no way to compare elements) only makes it harder to understand what's going on.

Reply:
I get what you mean, I just find the notion of having a `Top` signature that one can implement without even having a PO which would seem essential to being able to give such a thing non-helpful. But this is not a hill I am willing to die on!